[JENKINS-26660] Expose 'cleanSubmodules' functionality to Jenkins UI

slonopotamus opened the pull request:
[JENKINS-26660](https://issues.jenkins-ci.org/browse/JENKINS-26660) - add UI for cleaning of nested repositories/submodules

This PR exposes functionality created in jenkinsci/git-client-plugin#222 to Jenkins UI so it is possible to actually use it.

![image](https://user-images.githubusercontent.com/92637/69475939-4afcaf80-0de4-11ea-942e-13397ec4b244.png)

#449 attempted to do the same thing but was closed for various reasons.

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/jenkinsci/git-plugin/blob/master/CONTRIBUTING.md) doc
- [x] I have referenced the Jira issue related to my changes in one or more commit messages
- [x] I have added tests that verify my changes
- [x] Unit tests pass locally with my changes
- [x] I have added documentation as necessary
- [x] No Javadoc warnings were introduced with my changes
- [x] No spotbugs warnings were introduced with my changes
- [x] I have interactively tested my changes
- [x] Any dependent changes have been merged and published in upstream modules (like git-client-plugin)

## Types of changes

- [x] New feature (~non~-breaking change which adds functionality)

darxriggs commented on src/main/resources/hudson/plugins/git/extensions/impl/CleanBeforeCheckout/help-cleanSubmodules.html:
```diff
@@ -0,0 +1,3 @@
+<div>
+  Deletes untracked nested git repositories.<br/>
```
Is the `<br/>` required?

slonopotamus replied:
Removed it

darxriggs commented on src/main/resources/hudson/plugins/git/extensions/impl/CleanCheckout/help-cleanSubmodules.html:
```diff
@@ -0,0 +1,3 @@
+<div>
+  Deletes untracked nested git repositories.<br/>
```
Is the `<br/>` required?

slonopotamus replied:
Removed it

darxriggs commented on src/main/resources/hudson/plugins/git/extensions/impl/CleanBeforeCheckout/config.groovy:
```diff
@@ -0,0 +1,7 @@
+
+def f = namespace(lib.FormTagLib)
+
+f.entry(title: _("Clean submodules"), field: "cleanSubmodules") {
```
The question is how the entry should really be named in the UI. I think the help text describes it in a better way. That the git-client-plugin named the parameter `cleanSubmodules` is I think not that correct as cleaning submodules is just a subset of all potentially cleaned up `.git` folders.

MarkEWaite replied:
No objection from me if the option is more descriptive, as in

> Clean submodules and any other subdirectories which contain `.git` directories

slonopotamus replied:
Renamed option to "Delete untracked nested repositories"

darxriggs commented:
I think the `CleanAfterCheckoutTrait` and `CleanBeforeCheckoutTrait` also have to be updated.

MarkEWaite commented:
> I think using the `Warning.NONFINAL_FIELDS` suppression is the way to go. @MarkEWaite what's your thought on this?

Agreed though I don't see where there is a non-final field in the class.  What did I miss?

darxriggs commented:
There will be non-final fields when setters are used. This refers to
> Note that I did not use @DataBoundSetter how it was requested by @darxriggs in #449. The problem is with equalsContract test:

slonopotamus commented:
Updated PR, hopefully resolving all pointed issues. Also added tests for new property (it's now named `deleteUntrackedNestedRepositories`) to `GitSCMSourceTraitsTest`.

darxriggs commented on src/main/java/jenkins/plugins/git/traits/CleanAfterCheckoutTrait.java:
```diff
@@ -40,8 +41,8 @@
      * Stapler constructor.
      */
     @DataBoundConstructor
-    public CleanAfterCheckoutTrait() {
```
Hi @MarkEWaite, how strict do we have to be about backward compatibility for traits? Has to old constructor to be left in place too with a `@Deprecated` annotation and corresponding Javadoc (as it's done in many other classes)?

darxriggs replied:
I'll just answer myself. I think the old constructor should also be left in place. Otherwise e.g. existing Job DSL script may fail.

old way
```
multibranchPipelineJob('generated-git-traits-old') {
  branchSources {
    branchSource {
      source {
        git {
          id('1')
          remote('https://github.com/jenkinsci/job-dsl-plugin.git')
          traits {
            cleanBeforeCheckoutTrait() // <--- fails
            cleanAfterCheckoutTrait()  // <--- fails
          }
        }
      }
    }
  }
}
```

new way
```
multibranchPipelineJob('generated-git-traits-new') {
  branchSources { 
    branchSource { 
      source { 
        git { 
          id('2')
          remote('https://github.com/jenkinsci/job-dsl-plugin.git')
          traits { 
            cleanBeforeCheckoutTrait {
              extension {
                deleteUntrackedNestedRepositories(true)
              }
            }
            cleanAfterCheckoutTrait {
              extension {
                deleteUntrackedNestedRepositories(true)
              }
            }
          }
        }
      }
    }
  }
}
```

slonopotamus replied:
Do i need to deprecate old constructor? Do I need to mark both constructors with `@DataBoundConstructor` or only the new one? I'm not sure how all this stuff works.

darxriggs replied:
Only a single constructor is allowed to be annotated with `@DataBoundConstructor`.
Annotate the new one with `@DataBoundConstructor`.
Annotate the old one with `@Deprecated` and add Javadoc with a `@deprecated` description.
For an example see [UserMergeOptions](https://github.com/jenkinsci/git-plugin/blob/git-4.0.0/src/main/java/hudson/plugins/git/UserMergeOptions.java).

slonopotamus replied:
Okay, updated PR.

darxriggs replied:
I just tested the "old way" locally with Job DSL. Unfortunately it still fails. Job DSL seems to always use the `@DataBoundConstructor` annotated constructor. We have to investigate further.

The rest of the PR now looks good to me. Thanks for your effort.

slonopotamus replied:
I'm not sure what to do here. As I said, my Stapler (or whatever handles `@DataBoundConstructor`) knowledge is close to zero.

Hope that you can come up with a solution, then I'll implement it if such kind of backward-compatibility breakage is unacceptable.

As a last resort measure, we could create a _second_ trait (`cleanAfterCheckoutExtensionTrait` or smth like that), though it feels really awful.

darxriggs commented on src/test/java/jenkins/plugins/git/GitSCMSourceTraitsTest.java:
```diff
@@ -137,8 +137,18 @@ public void pimpped_out() throws Exception {
                                         hasProperty("localBranch", is("**"))
                                 )
                         ),
-                        Matchers.<SCMSourceTrait>instanceOf(CleanAfterCheckoutTrait.class),
```
Can you have a look if there are tests that cover reading old job configurations and maybe adapt/add such?
I am asking because it would be good to have tests that make sure that the "old" and "new" configuration XML (added property/setter) can be read properly (backward compatibility) for traits and extensions.

old
```
<extensions>
   <hudson.plugins.git.extensions.impl.CleanBeforeCheckout/>
</extensions>
```

new
```
<extensions>
   <hudson.plugins.git.extensions.impl.CleanBeforeCheckout>
      <deleteUntrackedNestedRepositories>true</deleteUntrackedNestedRepositories>
   </hudson.plugins.git.extensions.impl.CleanBeforeCheckout>
</extensions>
```

slonopotamus replied:
[This](https://github.com/jenkinsci/git-plugin/pull/792/commits/337e22ecbd77b228572cba9f51301873741a405a) is OK?

darxriggs replied:
This already looks ok. May I ask to you to additionally reduce the XML to the required minimum? So it's more obvious what this test is really about and it doesn't fail in cases that are not relevant for the backward compatibility of these two traits.

slonopotamus replied:
Reduced to:

```
<jenkins.plugins.git.GitSCMSource>
  <extensions>
    <hudson.plugins.git.extensions.impl.CleanCheckout />
    <hudson.plugins.git.extensions.impl.CleanBeforeCheckout />
  </extensions>
</jenkins.plugins.git.GitSCMSource>
```

MarkEWaite commented:
I wonder if we should consider leaving the traits without the extra clean option?  Traits are used in the multibranch pipeline context.  In that context, it is trivial for the pipeline author to insert their own call to `sh 'git clean -xffd'`.  That avoids compatibility concerns for the traits code and leaves this added clean option available for job types that don't use traits.

We don't need to keep traits perfectly aligned with the extensions, since there are already cases where extensions are not provided as traits.  That was generally because it was easier and clearer for the pipeline user to call a higher level function like `deleteDir()` than to have a 'wipe workspace' trait (and similar).

slonopotamus commented:
> I wonder if we should consider leaving the traits without the extra clean option?
I don't see how that's supposed to happen given that loading job from XML produces a _trait_.

MarkEWaite commented:
> > I wonder if we should consider leaving the traits without the extra clean option?
> 
> I don't see how that's supposed to happen given that loading job from XML produces a _trait_.

I thought that the traits were represented in the XML configuration file with the class name of the trait and that they were only used in contexts where traits were allowed.  If that is correct, that keeps them in the context of of multibranch pipelines and organization folders.

The idea was (and it may be the **wrong idea**) that the changes proposed in src/main/java/jenkins/plugins/git/traits/CleanBeforeCheckoutTrait.java would be reverted so that the extended `git clean` capability is not available through the traits at all.  That would avoid the compatibility concern (I think).

Jobs that use the traits would need to use `deleteDir()` or `sh 'git clean -xffd'` inside their pipeline scripts rather than adding a parameter to the CleanBeforeCheckoutTrait.

darxriggs commented:
The goal for me (maybe for @MarkEWaite too) should be to understand _extensions_ vs. _traits_ in more detail and come up with a suggestion.

In the next days I think I'll not have much time to look into this.

slonopotamus commented:
@darxriggs [This is how traits are related to extensions](https://github.com/jenkinsci/git-plugin/blob/master/src/main/java/jenkins/plugins/git/GitSCMSource.java#L221). So, even if we only want an extension, we still have to create a trait, because in runtime everything is done through traits. And trait needs a constructor that accepts extension, otherwise extension settings are lost.

I've updated PR. Now traits have argless `@DataBoundConstructor` constructor that will continue to work in pipelines and a constructor that accepts extension that is used when loading extensions from XML.

darxriggs commented:
@slonopotamus thanks for all your effort. Unfortunately the latest changes don't work as you expected. Try it yourself by running `mvn hpi:run` and create a multi-branch pipeline with the two affected behaviours added and the new option `enabled`. You will see in the serialized _job.xml_ that this value is ignored and `false` is stored.

slonopotamus commented:
Oh c'mooon! Turns out, `GitSCM` is saved in _completely_ dfferent ways for freestyle and multibranch projects.

This is freestyle (works properly with my changes):

```
<?xml version='1.1' encoding='UTF-8'?>
<project>
  <description></description>
  <keepDependencies>false</keepDependencies>
  <properties/>
  <scm class="hudson.plugins.git.GitSCM" plugin="git@4.0.1-SNAPSHOT">
    <configVersion>2</configVersion>
    <userRemoteConfigs>
      <hudson.plugins.git.UserRemoteConfig>
        <url>git://zzz@zzz.com</url>
      </hudson.plugins.git.UserRemoteConfig>
    </userRemoteConfigs>
    <branches>
      <hudson.plugins.git.BranchSpec>
        <name>*/master</name>
      </hudson.plugins.git.BranchSpec>
    </branches>
    <doGenerateSubmoduleConfigurations>false</doGenerateSubmoduleConfigurations>
    <submoduleCfg class="list"/>
    <extensions>
      <hudson.plugins.git.extensions.impl.CleanCheckout>
        <deleteUntrackedNestedRepositories>true</deleteUntrackedNestedRepositories>
      </hudson.plugins.git.extensions.impl.CleanCheckout>
      <hudson.plugins.git.extensions.impl.CleanBeforeCheckout>
        <deleteUntrackedNestedRepositories>true</deleteUntrackedNestedRepositories>
      </hudson.plugins.git.extensions.impl.CleanBeforeCheckout>
    </extensions>
  </scm>
  <canRoam>true</canRoam>
  <disabled>false</disabled>
  <blockBuildWhenDownstreamBuilding>false</blockBuildWhenDownstreamBuilding>
  <blockBuildWhenUpstreamBuilding>false</blockBuildWhenUpstreamBuilding>
  <triggers/>
  <concurrentBuild>false</concurrentBuild>
  <builders/>
  <publishers/>
  <buildWrappers/>
</project>
```

This is multibranch (loses `enabled` on save):

```
<?xml version='1.1' encoding='UTF-8'?>
<org.jenkinsci.plugins.workflow.multibranch.WorkflowMultiBranchProject plugin="workflow-multibranch@2.18">
  <actions/>
  <description></description>
  <properties/>
  <folderViews class="jenkins.branch.MultiBranchProjectViewHolder" plugin="branch-api@2.5.4">
    <owner class="org.jenkinsci.plugins.workflow.multibranch.WorkflowMultiBranchProject" reference="../.."/>
  </folderViews>
  <healthMetrics>
    <com.cloudbees.hudson.plugins.folder.health.WorstChildHealthMetric plugin="cloudbees-folder@6.9">
      <nonRecursive>false</nonRecursive>
    </com.cloudbees.hudson.plugins.folder.health.WorstChildHealthMetric>
  </healthMetrics>
  <icon class="jenkins.branch.MetadataActionFolderIcon" plugin="branch-api@2.5.4">
    <owner class="org.jenkinsci.plugins.workflow.multibranch.WorkflowMultiBranchProject" reference="../.."/>
  </icon>
  <orphanedItemStrategy class="com.cloudbees.hudson.plugins.folder.computed.DefaultOrphanedItemStrategy" plugin="cloudbees-folder@6.9">
    <pruneDeadBranches>true</pruneDeadBranches>
    <daysToKeep>-1</daysToKeep>
    <numToKeep>-1</numToKeep>
  </orphanedItemStrategy>
  <triggers/>
  <disabled>false</disabled>
  <sources class="jenkins.branch.MultiBranchProject$BranchSourceList" plugin="branch-api@2.5.4">
    <data>
      <jenkins.branch.BranchSource>
        <source class="jenkins.plugins.git.GitSCMSource" plugin="git@4.0.1-SNAPSHOT">
          <id>b77882ae-4976-4924-846d-c3e67691c2de</id>
          <remote></remote>
          <credentialsId></credentialsId>
          <traits>
            <jenkins.plugins.git.traits.BranchDiscoveryTrait/>
            <jenkins.plugins.git.traits.CleanAfterCheckoutTrait>
              <extension class="hudson.plugins.git.extensions.impl.CleanCheckout">
                <deleteUntrackedNestedRepositories>false</deleteUntrackedNestedRepositories>
              </extension>
            </jenkins.plugins.git.traits.CleanAfterCheckoutTrait>
            <jenkins.plugins.git.traits.CleanBeforeCheckoutTrait>
              <extension class="hudson.plugins.git.extensions.impl.CleanBeforeCheckout">
                <deleteUntrackedNestedRepositories>false</deleteUntrackedNestedRepositories>
              </extension>
            </jenkins.plugins.git.traits.CleanBeforeCheckoutTrait>
          </traits>
        </source>
        <strategy class="jenkins.branch.DefaultBranchPropertyStrategy">
          <properties class="empty-list"/>
        </strategy>
      </jenkins.branch.BranchSource>
    </data>
    <owner class="org.jenkinsci.plugins.workflow.multibranch.WorkflowMultiBranchProject" reference="../.."/>
  </sources>
  <factory class="org.jenkinsci.plugins.workflow.multibranch.WorkflowBranchProjectFactory">
    <owner class="org.jenkinsci.plugins.workflow.multibranch.WorkflowMultiBranchProject" reference="../.."/>
    <scriptPath>Jenkinsfile</scriptPath>
  </factory>
</org.jenkinsci.plugins.workflow.multibranch.WorkflowMultiBranchProject>
```

slonopotamus commented:
I reverted code to the version where traits have a `@DataBoundConstructor` that accepts extension object plus a deprecated no-arg constructor. I believe this is a cleanest variant of changes.

So everything works except pipelines that do

```
          traits {
            cleanBeforeCheckoutTrait() // <--- fails
            cleanAfterCheckoutTrait()  // <--- fails
          }
```
Dunno what to do with it further.

Aaarggh, never thought that adding a trivial boolean option would require so much effort.

darxriggs commented:
An option to keep backward compatibility could be as shown below.
a) Keep the old constructor annotated with `@DataBoundConstructor` for backward compatibility.
b) Add a new constructor with an `extension` parameter for future use.
c) Add a setter method annotated with `@DataBoundSetter` so the extension can be overwritten. Use it to transfer the value of the new option. It should not be annotated with `@Deprecated` otherwise Job DSL will report that `extension` is deprecated.

```
public class CleanBeforeCheckoutTrait extends GitSCMExtensionTrait<CleanBeforeCheckout> {

    /**
     * Stapler constructor.
     *
     * @deprecated use the new constructor that allows to set the extension.
     */
    @DataBoundConstructor
    public CleanBeforeCheckoutTrait() {
        this(new CleanBeforeCheckout());
    }

    public CleanBeforeCheckoutTrait(@NonNull CleanBeforeCheckout extension) {
        super(extension);
    }

    /**
     * @deprecated use the new constructor that allows to set the extension.
     */
    @DataBoundSetter
    public void setExtension(@NonNull CleanBeforeCheckout extension) {
        super.getExtension().setDeleteUntrackedNestedRepositories(extension.isDeleteUntrackedNestedRepositories());
    }

    ...
}
```

darxriggs commented:
@MarkEWaite do you think what I suggested could be the way to go or do you have any objections?

MarkEWaite commented:
@darxriggs I've never tried adding a constructor with more arguments than the DataBoundConstructor.  I'm not sure how it will behave.  I'm open to consider it, but don't have any experience which says that it will work in all the cases where it is expected to work.

@slonopotamus I am so sorry for the bumps and bruises and surprises hiding in this pull request.  Thanks for your patience as we learn together!

darxriggs commented:
The goal with this approach is that the new constructor could be used by end users who are using the plugin somehow programmatically with the next release. (I guess nobody will do so.) It will not be used by any implementation inside the plugin. And after some time remove the setter and no-arg constructor and put the `DataBoundConstructor` annotation to the new constructor. But I fear removing these deprecated artefacts might not happen.

If the chance that these deprecated artefacts will not be removed is pretty high or the effort and migration procedure is to high we could also just not add this extra constructor.

@MarkEWaite What's your opinion?

@slonopotamus Based on Mark's decision could you then implement the changes? I'd also want you to ask to add tests for reading the xml configuration of the old and new format for the `Traits` (similar as you already did for the `Extensions`). This could already be implemented before Mark's decision.

slonopotamus commented:
> I'd also want you to ask to add tests for reading the xml configuration of the old and new format for `<traits>` (similar as you already did for `<extensions>`)

[Done](https://github.com/jenkinsci/git-plugin/pull/792/files#diff-eead868e6822ece907ee90036703b044R86-R104). There are all 4 cases: old-style extension, old-style trait, new-style extension, new-style trait.

slonopotamus commented:
@darxriggs Do you have ideas how to write a test for trait usage in pipeline so that it would demonstrate regression we're talking about?

darxriggs commented:
I'll just get to have a look at earliest on Friday.

darxriggs commented on src/test/java/jenkins/plugins/git/GitSCMSourceTraitsTest.java:
```diff
@@ -83,6 +83,60 @@ public void modern() throws Exception {
     }
 
+    @Test
+    public void cleancheckout_v1() {
```
I would name this `cleancheckout_v1_extension` to make it more clear.

slonopotamus replied:
Done

darxriggs commented on src/test/resources/jenkins/plugins/git/GitSCMSourceTraitsTest/cleancheckout_v1_trait.xml:
```diff
@@ -0,0 +1,11 @@
+<jenkins.plugins.git.GitSCMSource>
+  <traits>
+    <jenkins.plugins.git.traits.BranchDiscoveryTrait/>
```
I guess the `BranchDiscoveryTrait` is left from an export. I would remove it as the tests are only about `Clean` traits as you also commented in the tests.

slonopotamus replied:
I'd need to do conditional logic in tests because `BranchDiscoveryTrait` is added implicitly when loading `<extensions>` but not when loading `<traits>`.

darxriggs replied:
Ok. Got it.

slonopotamus replied:
Ok, done.

darxriggs commented:
@slonopotamus if I got your question right it's about how to test the Job DSL stuff.

You can have a look at the the tests of the plugin itself - e.g. https://github.com/jenkinsci/job-dsl-plugin/blob/job-dsl-1.76/job-dsl-plugin/src/test/groovy/javaposse/jobdsl/plugin/JenkinsJobManagementSpec.groovy#L524-L526

What I currently can think of:
Add the job-dsl plugin as test dependency to _pom.xml_.
Create two other test methods like `cleancheckout_v1_jobdsl` and `cleancheckout_v2_jobdsl`
```
String jobDslScript = "..."
FreeStyleProject seedJob = jenkinsRule.createFreeStyleProject('seed')
seedJob.buildersList.add(new ExecuteDslScripts(jobDslScript))
jenkinsRule.buildAndAssertSuccess(seedJob)
// verification of created job goes here
```
The `jobDslScript` can contain something very similar to my https://github.com/jenkinsci/git-plugin/pull/792#discussion_r349876573.
This will create a job and the generated xml can be verified similar to the existing tests.

slonopotamus commented:
I [tried](https://github.com/slonopotamus/git-plugin/commit/23ba946ac52f796861b0b80e8f931e191b12a422) writing a test for Job DSL. The problem is that Job DSL methods [require plugins to be installed](https://github.com/jenkinsci/job-dsl-plugin/blob/master/job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/DslFactory.groovy#L138):

```
ERROR: (script, line 1) version 2.10 or later of plugin 'workflow-multibranch' needs to be installed
```

There's one more minor annoyance: `ExecuteDslScripts` constructor isn't public, so I have to hack around this in test.

And while relevant classes are all on classpath when we're running tests, Jenkins doesn't think that it has any plugins. `@WithPlugin` jenkins-test-harness test annonation is a no-go, because we'll end up with multiple versions of everything (including git-plugn itself!)- one on test classpath and one from plugin.

On the other side, seems like [jobdsl doesn't care as much about backward compatibility as you do](https://github.com/jenkinsci/job-dsl-plugin/commit/afcc6869c14aba348cc8f4c21a6cdb3e04e541e9).

darxriggs commented:
I added some comments to your JobDSL test commit just to provide some more information.
I think it's fine not to spend more effort on trying to find out how to write tests for JobDSL compatibility.

darxriggs commented:
So we just have https://github.com/jenkinsci/git-plugin/pull/792#issuecomment-558492249 to resolve. 

Any feedback from you @MarkEWaite? Which option do we want to go? 

**Option A) Break at least JobDSL compatibility**
Then we can use the current implementation.

**Option B) Keep compatibility 100%**
Then we should have trait implementations that only add a new setter.
(Let's skip the the 2nd constructor as I originally suggested.) 
```
public class CleanBeforeCheckoutTrait extends GitSCMExtensionTrait<CleanBeforeCheckout> {
    ...
    @DataBoundSetter
    public void setExtension(@NonNull CleanBeforeCheckout extension) {
        super.getExtension().setDeleteUntrackedNestedRepositories(extension.isDeleteUntrackedNestedRepositories());
    }
    ...
}
```

BTW: The traits have been implemented originally with an `extension` parameter which was then removed later - see https://github.com/jenkinsci/git-plugin/commit/f09e7a299a0ca5ebd1f8492c29255f934ef61d27#diff-67eee29d51b821593427e7401dd837b6.

MarkEWaite commented:
I prefer option A with the current implementation

darxriggs commented:
I performed also some manual tests. Everything looks good to me so far.

@slonopotamus Thanks for your patience and effort.

@MarkEWaite Can you have a final view on it and merge it?

nvuillam commented:
Guys :) 

If confirm that the following code fails, and we had to spend some hours to find this topic, which is the only one containing the solution :)
Please think about ascending compatibility next time ^^

```
            cleanBeforeCheckoutTrait() // <--- fails
            cleanAfterCheckoutTrait()  // <--- fails
```

slonopotamus commented:
@nvuillam 

1. Where did you read about `cleanAfterCheckoutTrait()` from? Let's update that document with correct syntax.
2. If you read this PR carefully, you'll see that we _did_ spend nontrivial amount of time thinking about compatibility.

MarkEWaite commented:
I'll need to ask the Job DSL maintainers for specific guidance on what we should have done in that case to maintain compatibility.  We could have skipped the addition of submodule cleaning to the trait, but then I think that would have prevented its use in Job DSL.

The git plugin documentation does not include Pipeline samples or JobDSL samples. This may indicate that it needs to include those samples so that we can detect changes that affect other plugins.

darxriggs commented:
@MarkEWaite You could add a comment on the release page (if possible with release drafter in place) that this change breaks Job DSL backward compatibility and the hint about the new syntax https://github.com/jenkinsci/git-plugin/pull/792#discussion_r349876573.

MarkEWaite commented:
I had added it to the release notes yesterday with a link to a different comment in the pull request.  Will update it today with the link to the [comment you recommended](https://github.com/jenkinsci/git-plugin/pull/792#issuecomment-565757445) and a link to the [comment in the bug report](https://github.com/jenkinsci/git-plugin/pull/792#issuecomment-565757445).
